### PR TITLE
Upgrade EVM to Shanghai

### LINKED ISSUE #5959: Upgrade EVM to support Shanghai
### Problem

Ethereum Mainnet will be upgrading to Shanghai on 12 April, which includes changes to the EVM we will want to support, such as the PUSH0 operation.

### Solution

Update the EVM library to a more current version of Hyperledger Besu, such as 23.1.2.

### Alternatives

Alternative is to not upgrade and become incompatible with future versions of solidity.  Not a good option.

### PR DESCRIPTION
**Pull request by shemnon**
**Description**:

Upgrade Hyperledger Besu to 23.1.2 and wire in activation of the Shanghai EVM upgrades to v0.38

**Related issue(s)**:

Fixes #5959

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [X] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/HederaEvmWorldStateTokenAccount.java [hunk 1]
```diff
@@ -46,7 +46,7 @@ public HederaEvmWorldStateTokenAccount(final Address address) {
 
     @Override
     public Bytes getCode() {
-        return interpolatedCode().getContainerBytes();
+        return interpolatedCode().getBytes();
```
Comment (Nana-EC) [inline hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/HederaEvmWorldStateTokenAccount.java:49]: Q: why the change? Should probably be in PR description details

File: hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/contracts/operations/AbstractEvmRecordingCreateOperation.java [hunk 2]
```diff
@@ -151,7 +149,7 @@ private void spawnChildMessage(final MessageFrame frame) {
                 .sender(frame.getRecipientAddress())
                 .value(value)
                 .apparentValue(value)
-                .code(CodeFactory.createCode(inputData, Hash.EMPTY, 0, false))
+                .code(CodeFactory.createCode(inputData, 0, false))
```
Comment (Nana-EC) [inline hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/contracts/operations/AbstractEvmRecordingCreateOperation.java:152]: Q: why the change? Besu no longer needs the hash here I guess.
Should probably be in PR description details

File: hedera-node/hedera-evm/src/test/java/com/hedera/node/app/service/evm/contracts/operations/HederaEvmCreate2OperationTest.java [hunk 3]
```diff
@@ -88,7 +88,7 @@ void enabledOnlyIfCreate2IsEnabled() {
     void computesExpectedTargetAddress() {
         final var expectedAddress = Address.BLS12_G1ADD;
         final var canonicalSource = Address.BLS12_G1MULTIEXP;
-        final var besuOp = new Create2Operation(gasCalculator);
+        final var besuOp = new Create2Operation(gasCalculator, 48 * 1024);
```
Comment (Nana-EC) [inline hedera-node/hedera-evm/src/test/java/com/hedera/node/app/service/evm/contracts/operations/HederaEvmCreate2OperationTest.java:91]: Q: Is this 48kb? WHat's the significance of the choice?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/ContractsModule.java [hunk 4]
```diff
@@ -210,6 +231,15 @@ static ContractCreationProcessor provideV_0_34ContractCreateProcessor(
         return new ContractCreationProcessor(gasCalculator, evm, true, List.copyOf(validationRules), 1);
     }
 
+    @Provides
+    @Singleton
+    @IntoMap
+    @StringKey(ContractsV_0_38Module.EVM_VERSION_0_38)
+    static ContractCreationProcessor provideV_0_38ContractCreateProcessor(
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/ContractsModule.java:238]: Q: So even when the details are exactly the same we still have to add a new method for each new version?
Is that because of the lookup logic based on the key above?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/operation/HederaSStoreOperation.java [hunk 5]
```diff
@@ -58,13 +60,15 @@ public OperationResult execute(final MessageFrame frame, final EVM evm) {
...
+        final Supplier<UInt256> oldValue = Suppliers.memoize(() -> account.getStorageValue(key));
+        final Supplier<UInt256> originalValue = Suppliers.memoize(() -> account.getOriginalStorageValue(key));
         if (account == null) {
             return ILLEGAL_STATE_CHANGE_RESULT;
         }
         final var address = account.getAddress();
         final var slotIsWarm = frame.warmUpStorage(address, key);
         final var calculator = gasCalculator();
-        final var calcGasCost = calculator.calculateStorageCost(account, key, value)
+        final var calcGasCost = calculator.calculateStorageCost(value, oldValue, originalValue)
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/operation/HederaSStoreOperation.java:71]: Maybe some comments to inform how we're to understand the 3 separate but similarly name variables.

File: hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.sol [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    function opExtCodeHash(address _addr) public view returns (bytes32 _resp){
+        assembly {
+            _resp := extcodehash(_addr)
+        }
+    }
+
+    function opPush0() public pure returns (uint _resp) {
+        assembly {
+        _resp := add(msize(), 0x5f)
+//      _resp := add(verbatim_0i_1o(hex"5f"), 0x5f)
```
Comment (Nana-EC) [inline hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.sol:40]: nit: remove 
nit2: new line at end of file :)

File: hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.yul [hunk 7]
```diff
@@ -0,0 +1,347 @@
+/// @use-src 0:"OpcodesContract.sol"
+object "NewOpcodes_51" {
```
Comment (Nana-EC) [inline hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.yul:2]: I think this is the first `*.yul` file added right?
Going forward what's the suggestion do you now have to add a `.sol` & a `.yul`, also how should it be utilized?

**Review by Nana-EC — COMMENTED** [review]
Nice. Some questions and comments

File: hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/HederaEvmWorldStateTokenAccount.java (hunk 1 above)
Comment (shemnon) [reply hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/HederaEvmWorldStateTokenAccount.java:49]: The API provided by Besu changed.

File: hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/contracts/operations/AbstractEvmRecordingCreateOperation.java (hunk 2 above)
Comment (shemnon) [reply hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/contracts/operations/AbstractEvmRecordingCreateOperation.java:152]: The API provided by besu changed.

File: hedera-node/hedera-evm/src/test/java/com/hedera/node/app/service/evm/contracts/operations/HederaEvmCreate2OperationTest.java (hunk 3 above)
Comment (shemnon) [reply hedera-node/hedera-evm/src/test/java/com/hedera/node/app/service/evm/contracts/operations/HederaEvmCreate2OperationTest.java:91]: (a) the API changed, so we needed some value
(b) 48k is defined in EIP-3860
(c) it's test code, this isn't used but we need some value for the constructor.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/ContractsModule.java (hunk 4 above)
Comment (shemnon) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/ContractsModule.java:238]: This is how it needs to work to support dagger.  Because we inject a different version of the EVM (marked with the @V_0_38 annotation) we need a whole new provides definition.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/operation/HederaSStoreOperation.java (hunk 5 above)
Comment (shemnon) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/operation/HederaSStoreOperation.java:71]: The API from besu changed.  It's explained in the besu code already.

File: hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.sol (hunk 6 above)
Comment (shemnon) [reply hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.sol:40]: No, this should not be removed.  It is an interim step because (a) solidity doesn't support the push0 opcode in assembly yet and (b) doesn't support the verbatim in solidity.  So this is a placeholder to show what should be added.

NewOpcodes.sol (from which this is derived) didn't have a newline at the end of a file.

File: hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.yul (hunk 7 above)
Comment (shemnon) [reply hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.yul:2]: This is because Solidity doen't support PUSH0 yet in assembly or YUL.  Building instructions are at the top of the .sol file.

**Review by Nana-EC — APPROVED** [review]

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                        .payingWith(BOB)
+                        .via(pushResult)
+                        .logged()))
+                .then(getTxnRecord(pushResult)
+                        .hasPriority(recordWith()
+                                .contractCallResult(resultWith()
+                                        .resultViaFunctionName(
+                                                OP_PUSH_ZERO,
+                                                pushZeroContract,
+                                                isLiteralResult((new Object[] {BigInteger.valueOf(0x5f)})))))
```
Comment (david-bakin-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java:91]: I obviously don't know how this test suite works but this seems odd to me: Why is the literal result of a `PUSH0` test a `0x5F` instead of a `0x00`?  How does this test work?  Oh wait, I see now `0x5F` is the opcode for `PUSH0` - but then why doesn't this test actually test a `0` was pushed on the stack?  Or is that not the purpose of this test - opcode correct execution is tested elsewhere (like in besu itself) and this test suite is ... just to make sure the opcode actually exists?

(I suppose it is too much to ask but is there a doc somewhere that describes what these tests under `bdd.suites.contract` are for?  I'm not asking you to write one now, but did anyone write one ever?  Where is it?)

File: hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.yul (hunk 7 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.yul:2]: Perhaps add that first sentence as an additional comment up at the top of this file with the building instructions?  It's kind of a `TODO` isn't it?  When Solidity gets versioned we can take it and fix this?

**Review by david-bakin-sl — DISMISSED** [review]
Some questions - but only minor things about the code itself.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/ContractsV_0_38Module.java [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+@Module
+public interface ContractsV_0_38Module {
+
+    String EVM_VERSION_0_38 = "v0.38";
+
+    @Provides
+    @Singleton
+    @V_0_38
+    static BiPredicate<Address, MessageFrame> provideAddressValidator(
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/ContractsV_0_38Module.java:72]: This was pre-existing work, as you said it was already duplicated in v0.30 and v0.34.  

We could factor it out short term, but it will come back just as soon as we have versions with different precompiled contract sets.  The duplication is a emergent property of using dagger.

Delegating it to another (shared) method would further erode the readability of these injectors.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java (hunk 8 above)
Comment (shemnon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java:91]: There is a whole lot of yak shaving that goes on in an EET.  I can't change that, the alternative is to not test that PUSH0 is activated in 0.38 and not activated in 0.34.

Per the conventions the solidity is kept in a different file, (referenced via `pushZeroContract` variable, via static variable `CONTRACT`, going to `"OpcodesContract"`, which resolves to the [OpcodeContract](https://github.com/hashgraph/hedera-services/tree/4a89d5a32a7292709f79f5509c6e26d2cb20ccc9/hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract) directory in the resources fork of the test fork of the test-clients module.

The `OP_PUSH_ZERO` static field maps to `"opPush0"` which corresponds to the [`opPush0`](https://github.com/hashgraph/hedera-services/blob/4a89d5a32a7292709f79f5509c6e26d2cb20ccc9/hedera-node/test-clients/src/main/resource/contract/contracts/OpcodesContract/OpcodesContract.sol#L37) function in the corresponding smart contract.  I'll address how this is compiled in another comment, but the commented out line is actually what gets executed: a push0, a push1  pushing 0x5f, then an add, then returning that value.  A push 0 returning a zero could be mistaken for a do-nothing function success, so I had to do something.

This test is duplicated twice (this function and the one below), with different values in the `overriding` section in the when clause.  One activating the 0.38 VM and one the 0.34.  The assert results are different because the VM code should revert if not activated.

File: buildSrc/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts [hunk 10]
```diff
@@ -46,6 +46,7 @@ extraJavaModuleInfo {
 
     automaticModule("org.hyperledger.besu:secp256k1", "org.hyperledger.besu.secp256k1")
     automaticModule("org.hyperledger.besu.internal:crypto", "org.hyperledger.besu.crypto")
+    automaticModule("tech.pegasys:jc-kzg-4844", "tech.pegasys.jckzg4844")
```
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply buildSrc/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts:49]: I get that we take this module or play a different game.  But - put aside the fact you've been steeped in EVM+BESU for the last year(s).  Take a good look at the string **"jdkzj4844"**.  Remember: you don't know what 4844 is in Ethereum-space.  Now, doesn't that look like something on your file system you'd find malware in?  Just a comment, Danno, just a comment ...

[the remark this answers is not in the document]
Comment (shemnon) [reply buildSrc/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts:49]: No other entries in that file have comments, this file isn't a descriptor of libraries but a mapping for auto-modularization.  A comment would be out of place.

[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply buildSrc/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts:49]: Oh.  Out of place.  Ok, how about putting it in `com.hedera.hashgraph.conventions.gradle.kts` - there's already a comment in there, one more won't break the parser.

People are going to be wondering why the heck we're pulling a module from some random place on the internet.  They won't know that Consensus consulted the stars before pulling `pegasys` out of thin air.

**Review by nathanklick — APPROVED** [review]

**Review by lukelee-sl — APPROVED** [review]

**Review by nickpoorman — APPROVED** [review]

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    List<HapiSpec> positiveSpecs() {
+        return List.of(pushZeroHappyPathWorks(), pushZeroDisabledInV034());
+    }
+
+    private HapiSpec pushZeroHappyPathWorks() {
+        final var pushZeroContract = CONTRACT;
+        final var pushResult = "pushResult";
+        return defaultHapiSpec("prngPrecompileHappyPathWorks")
```
Comment (tinker-michaelj) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java:73]: Typo here - `"PushZeroHappyPathWorks"`

**Review by tinker-michaelj — APPROVED** [review]
LGTM, tyvm @shemnon 🥂

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java (hunk 11 above)
Comment (Nana-EC) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java:73]: Resolving to unblock but we can do this in follow up

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java (hunk 8 above)
Comment (Nana-EC) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/PushZeroOperationSuite.java:91]: Resolving to unblock. 
We should consider adding comments in follow up

File: buildSrc/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply buildSrc/src/main/kotlin/com.hedera.hashgraph.jpms-modules.gradle.kts:49]: Resolving to unblock. Let's see what improved comments and documentation we can add to the repo or PR in future for changes like this where possible.
